### PR TITLE
sjawhar-legion-239: add learnings tracking to handoff schema and workflows

### LINKED ISSUE #239: Add learnings tracking to handoff schema and workflows
## Parent

Part of #218 (Knowledge compounding). Depends on #238 (knowledge injection to all phases).

## Problem

When workers inject learnings from `docs/solutions/`, there's no feedback signal. We don't know which learnings were injected into each phase or whether they were helpful. Without this data, we can't evolve the knowledge base (#240).

Currently only `PlanHandoff` has a `learningsUsed: string[]` field. Other phases have no learnings tracking. The `learningsUsed` name is inconsistent with the proposed `learningsInjected`/`learningsHelpful` vocabulary.

## Solution

Extend the handoff schema to track learnings injection and helpfulness across all phases. Migrate from `learningsUsed` to the new field names. Add recording steps to each workflow.

### Terminology

- **`learningsInjected`**: Canonical `docs/solutions/` file paths of learnings that were presented to the worker at the start of the phase. "Injected" = "consulted" — if the injection step ran and output learnings, the worker received them. No separate "consulted" field needed.
- **`learningsHelpful`**: Subset of `learningsInjected` that the worker explicitly marked as having materially helped this phase's output. Empty array means none were helpful (not a failure signal — the learnings may simply not have been relevant to this specific issue).

Both fields use canonical `docs/solutions/` relative file paths as identifiers (e.g., `daemon/controller-lifecycle-separation.md`). These stable IDs enable aggregation across issues in #240.

### Design

**1. Schema changes (`packages/daemon/src/handoff/types.ts` and `schema.ts`):**

Add two new optional fields to `BaseHandoff`:

```typescript
// In BaseHandoff — available to all phases
learningsInjected?: string[];  // Canonical docs/solutions/ paths injected into this phase
learningsHelpful?: string[];   // Subset the worker found materially helpful
```

**2. Migrate `learningsUsed` in PlanHandoff:**

Remove `learningsUsed` from `PlanHandoff` interface and `planSchema`. The new `learningsInjected` on `BaseHandoff` replaces it. For read compatibility:
- Add a `transform` step to the plan schema that maps `learningsUsed` → `learningsInjected` when reading old handoff data
- All new writes use only `learningsInjected`

This avoids split vocabulary — there's exactly one canonical field name.

**3. Zod schema changes (`schema.ts`):**

Add to `baseHandoffSchema`:
```typescript
learningsInjected: z.array(z.string()).optional(),
learningsHelpful: z.array(z.string()).optional(),
```

Add backward-compat transform to plan schema:
```typescript
const planSchema = baseHandoffSchema.extend({
  // ... existing fields ...
  learningsUsed: z.array(z.string()).optional(), // deprecated, read-only compat
}).transform((data) => {
  if (data.learningsUsed && !data.learningsInjected) {
    return { ...data, learningsInjected: data.learningsUsed, learningsUsed: undefined };
  }
  return data;
});
```

**4. Workflow handoff write updates:**

Each workflow's handoff write section adds the new fields. Workers populate `learningsInjected` from whatever the injection step (#238) output. Example for implement phase:

```bash
legion handoff write --phase implement --workspace . <<'HANDOFF' 2>/dev/null || true
{
  "filesChanged": ["src/auth/middleware.ts", "src/auth/types.ts"],
  "trickyParts": ["Token refresh race condition"],
  "learningsInjected": ["daemon/controller-lifecycle-separation.md", "architecture-patterns/shared-state-file-ownership.md"],
  "learningsHelpful": ["daemon/controller-lifecycle-separation.md"]
}
HANDOFF
```

**5. Helpfulness assessment step:**

Add a step at the end of each workflow (before handoff write) where the worker reviews which injected learnings were actually relevant:

> "Review the learnings injected at Step 1.5. For each, assess: did this learning materially influence your work, prevent a mistake, or provide useful context for this phase? List only those canonical paths in `learningsHelpful`. If none were helpful, use an empty array. If no learnings were injected, omit both fields from handoff."

This goes in all five workflows: architect.md, plan.md, implement.md, test.md, review.md — before their respective handoff write steps.

**6. Update plan.md handoff write example:**

Replace `learningsUsed` with `learningsInjected` in the plan.md step 5.5 handoff write example.

## Acceptance Criteria

- [ ] `BaseHandoff` interface in `types.ts` includes `learningsInjected?: string[]` and `learningsHelpful?: string[]`
- [ ] `baseHandoffSchema` in `schema.ts` includes both new fields as `z.array(z.string()).optional()`
- [ ] `PlanHandoff.learningsUsed` is removed from the interface; plan schema includes a backward-compat transform mapping `learningsUsed` → `learningsInjected`
- [ ] All five workflow handoff write examples (architect, plan, implement, test, review) include `learningsInjected` and `learningsHelpful` fields using canonical `docs/solutions/` relative paths
- [ ] Each workflow has a helpfulness assessment step before its handoff write, with the exact assessment prompt specified above
- [ ] plan.md handoff write example uses `learningsInjected` (not `learningsUsed`)
- [ ] Existing handoff tests pass (backward compat: old plan handoff with `learningsUsed` still parses correctly and maps to `learningsInjected`)
- [ ] New test: plan handoff with `learningsUsed` validates and transforms to `learningsInjected`
- [ ] New test: handoff data with both `learningsInjected` and `learningsHelpful` validates for each phase
- [ ] New test: handoff data without learnings fields still validates (optional fields)
- [ ] `bunx tsc --noEmit` passes
- [ ] `bun test` passes (including handoff tests)

## Testing Infrastructure

**Available:**
- `bun test` runs handoff schema tests at `packages/daemon/src/cli/__tests__/handoff.test.ts`
- `bunx tsc --noEmit` for type checking

**Gaps:** None — existing test infrastructure covers this.

## Notes for Implementation

- `BaseHandoff` changes propagate to all phases via inheritance — no per-phase schema additions needed
- The `learningsUsed` → `learningsInjected` migration must maintain read compat for existing handoff data on branches
- Workers populate `learningsInjected` from the injection step output (#238)
- Workers populate `learningsHelpful` from their own assessment (subset of `learningsInjected`)
- If no learnings were injected (injection step found no matches), omit both fields entirely
- Empty `learningsHelpful: []` is valid — means learnings were injected but none were relevant

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #239

## Summary

Add `learningsInjected` and `learningsHelpful` fields to the handoff schema so workers can record which learnings from `docs/solutions/` were presented and which were actually useful.

### Changes

- **`BaseHandoff` interface** (`types.ts`): Added `learningsInjected?: string[]` and `learningsHelpful?: string[]` — available to all phases via inheritance
- **`baseHandoffSchema`** (`schema.ts`): Added both fields as `z.array(z.string()).optional()`
- **`PlanHandoff.learningsUsed` migration**: Removed from interface; plan schema retains `learningsUsed` for read compatibility with a runtime transform in `validatePhaseHandoff` that maps `learningsUsed` → `learningsInjected`
- **All 5 workflow handoff write examples** (architect, plan, implement, test, review): Updated to include both new fields with canonical `docs/solutions/` paths
- **Helpfulness assessment step**: Added to all 5 workflows before their handoff write step — workers review injected learnings and identify which were materially helpful
- **`review.md`**: Updated reference from `plan.learningsUsed` to `plan.learningsInjected`
- **Tests**: Added 4 new test cases covering backward compat migration, field precedence, per-phase validation, and optional field handling

### Verification

- `bunx tsc --noEmit` — passes
- `bun test` — all 24 handoff tests pass (14 ledger + 10 CLI)
- `bunx biome check` — passes

### CODE REVIEW

**Comment by sjawhar** [comment]
## Behavioral Test Results

### Summary
**PASS** — 12/12 acceptance criteria verified

### Results
| Criterion | Status | Evidence |
|-----------|--------|----------|
| `BaseHandoff` interface includes `learningsInjected?: string[]` and `learningsHelpful?: string[]` | ✅ | Verified by code inspection: `types.ts:22-24` — both fields present with correct JSDoc comments |
| `baseHandoffSchema` includes both fields as `z.array(z.string()).optional()` | ✅ | Verified by code inspection: `schema.ts:44-45` — exact Zod declarations match spec |
| `PlanHandoff.learningsUsed` removed from interface; plan schema has backward-compat transform | ✅ | `types.ts:36-44` — no `learningsUsed` in interface. `schema.ts:72-73` — deprecated field retained for read compat. `schema.ts:146-153` — runtime transform in `validatePhaseHandoff` migrates `learningsUsed` → `learningsInjected`. Deviation from spec (transform in function not schema `.transform()`) — justified by Zod discriminatedUnion limitation documented in implementer handoff |
| All 5 workflow handoff write examples include `learningsInjected` and `learningsHelpful` | ✅ | Grep across all 5 workflows confirmed: `architect.md:119-120`, `plan.md:371-372`, `implement.md:366-367` (fresh) + `514-515` (address-comments), `test.md:270-271`, `review.md:202-203` |
| Each workflow has helpfulness assessment step before handoff write | ✅ | All 5 workflows contain the exact assessment prompt blockquote: `architect.md:104`, `plan.md:357`, `implement.md:356` + `504`, `test.md:260`, `review.md:189` |
| `plan.md` handoff write uses `learningsInjected` (not `learningsUsed`) | ✅ | `plan.md:371` — uses `learningsInjected`. No `learningsUsed` references in any workflow handoff write |
| Existing handoff tests pass (backward compat) | ✅ | `bun test packages/daemon/src/handoff/` — 14/14 pass, 0 fail, 55 expect() calls |
| New test: plan handoff `learningsUsed` → `learningsInjected` migration | ✅ | `ledger.test.ts:260-291` — writes old-style plan JSON with `learningsUsed`, verifies `readPhaseHandoff` returns `learningsInjected` and removes `learningsUsed` |
| New test: `learningsInjected`+`learningsHelpful` validates for each phase | ✅ | `ledger.test.ts:321-350` — tests architect, plan, implement, test, review phases with both fields |
| New test: handoff without learnings fields validates (optional) | ✅ | `ledger.test.ts:352-365` — confirms omitting both fields still validates |
| `bunx tsc --noEmit` passes | ✅ | Only output is pre-existing TS5101 deprecation warning about `baseUrl` in tsconfig — no type errors |
| `bun test` passes | ✅ | 954 pass, 3 fail (pre-existing: 1 daemon entry test + 2 pulumi module errors). All 14 handoff tests pass. |

### Additional Verification: CLI End-to-End
| Test | Status | Evidence |
|------|--------|----------|
| Write handoff with learnings fields via CLI | ✅ | `legion handoff write --phase test` with `learningsInjected` + `learningsHelpful` → read back shows both fields preserved correctly with auto-populated `schemaVersion`, `phase`, `completed` |
| Write handoff without learnings fields (optional) | ✅ | `legion handoff write --phase implement` without learnings → validates, fields absent as expected |
| TypeScript API backward compat migration | ✅ | Direct `bun -e` test: `readPhaseHandoff` and `validatePhaseHandoff` correctly migrate `learningsUsed` → `learningsInjected` and remove `learningsUsed` from output |
| Field precedence test (both `learningsUsed` + `learningsInjected`) | ✅ | `ledger.test.ts:293-319` — when both exist, `learningsInjected` preserved, `learningsUsed` not overwritten |

### Test Quality Critique
- **Tests are meaningful, not decorative.** The 4 new test cases cover: backward compat migration, field precedence, per-phase validation loop, and optional field handling.
- **No excessive mocking.** Tests use real filesystem operations (mkdtemp, writeFile, readFile) — no mocking of the validation or ledger layer.
- **Edge cases covered:** Tests verify corrupt JSON handling, schema mismatches, non-ISO timestamps, wrong field types, and reserved field overwrite protection (pre-existing tests).
- **Minor gap (P3):** No test for `learningsHelpful` containing values NOT in `learningsInjected` (invalid subset). This is a validation gap — the schema allows it since both are independent `z.array(z.string()).optional()`. The issue spec says `learningsHelpful` is a "subset" but this is not enforced by schema validation. Worth noting but not blocking.

### Documentation Feedback
- Workflow files are clear and self-documenting — the helpfulness assessment prompt is a well-written blockquote that workers can follow directly.
- Field descriptions in each workflow are consistent across all 5 files.
- The `@deprecated` JSDoc on `learningsUsed` in `schema.ts:72` provides clear migration guidance.
- `review.md:74` correctly references `plan.learningsInjected` (updated from `learningsUsed`).

### Observations
- The `validatePhaseHandoff` transform approach (instead of Zod `.transform()`) is a valid workaround for the `z.discriminatedUnion` limitation — documented in implementer handoff `trickyParts`.
- The compiled `legion` CLI binary does not reflect source changes (expected — needs rebuild). All validation was done against source code via `bun test` and direct TypeScript API calls.
- Pre-existing test failures (3) are unrelated: 1 daemon entry env var test + 2 pulumi module import errors.
- Biome check passes (4 files, no issues).

### Config Compliance
- require_specific_task: false — N/A (no repo config)
- require_taiga_evidence: false — N/A (no repo config)

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 2 suggestions

**Verdict: Approved to merge.** All 12 acceptance criteria are met. Schema changes are correct, backward-compat migration works, all five workflows updated consistently, and test coverage is solid.

---

### Acceptance Criteria Verification

| # | Criterion | Status |
|---|-----------|--------|
| 1 | `BaseHandoff` has `learningsInjected?: string[]` and `learningsHelpful?: string[]` | ✅ types.ts lines 21-24 |
| 2 | `baseHandoffSchema` includes both as `z.array(z.string()).optional()` | ✅ schema.ts lines 44-45 |
| 3 | `PlanHandoff.learningsUsed` removed from interface; backward-compat migration | ✅ types.ts + schema.ts validatePhaseHandoff |
| 4 | All 5 workflow handoff write examples include both fields | ✅ architect, plan, implement (×2), test, review |
| 5 | Each workflow has helpfulness assessment step before handoff write | ✅ All 5 workflows, correct step references |
| 6 | plan.md uses `learningsInjected` (not `learningsUsed`) | ✅ |
| 7 | Existing handoff tests pass | ✅ CI green |
| 8 | New test: `learningsUsed` → `learningsInjected` migration | ✅ |
| 9 | New test: both fields validate per phase | ✅ 5 phases tested |
| 10 | New test: optional fields still validate | ✅ |
| 11 | `bunx tsc --noEmit` passes | ✅ CI green |
| 12 | `bun test` passes | ✅ CI green (24 handoff tests) |

### Key Design Decision

The issue spec suggested using `.transform()` on `planSchema` for the `learningsUsed` → `learningsInjected` migration. The implementer correctly identified that `z.discriminatedUnion` requires `ZodObject` members (not `ZodEffects` from `.transform()`), and placed the migration logic in `validatePhaseHandoff` instead. This achieves the same result and is the right call.

### P3 Findings (non-blocking)

1. **Test gap — `learningsUsed` removal in precedence test**: The "preserves learningsInjected when both exist" test verifies `learningsInjected` has the correct value but doesn't assert `"learningsUsed" in handoff` is false. The migration test does check this, but mirroring it here would make the precedence test self-contained.

2. **`retro` phase not in per-phase learnings test**: The per-phase validation test covers architect/plan/implement/test/review but not retro. Since retro inherits from `BaseHandoff` and the issue doesn't require retro learnings tracking, this is expected — but noting for completeness.

### Evidence
- CI status: lint ✅ (6s), test ✅ (39s), typecheck ✅ (17s)
- Tester handoff: 12 passed, 0 failed
- No bot review comments on PR

File: packages/daemon/src/handoff/__tests__/ledger.test.ts [hunk 1]
```diff
@@ -256,4 +256,111 @@ describe("handoff ledger", () => {
...
+  it("validates handoff data with learningsInjected and learningsHelpful for each phase", async () => {
+    workspaceDir = await mkdtemp(path.join(os.tmpdir(), "legion-handoff-"));
+
+    const phases = [
+      { phase: "architect" as const, extra: { scope: "small" } },
+      { phase: "plan" as const, extra: { taskCount: 3 } },
+      { phase: "implement" as const, extra: { filesChanged: ["a.ts"] } },
+      { phase: "test" as const, extra: { passed: 5, failed: 0 } },
+      { phase: "review" as const, extra: { verdict: "approved" } },
+    ];
```
Comment (sjawhar) [inline packages/daemon/src/handoff/__tests__/ledger.test.ts]: **[P3 — Minor]:** The per-phase loop covers architect/plan/implement/test/review but not `retro`. Since `retro` inherits `learningsInjected`/`learningsHelpful` from `BaseHandoff` via the schema, it would validate — but it's not exercised here. Not blocking since the issue doesn't require retro learnings tracking.
